File: app/src/main/java/com/test/upcoming/ui/activity/detail/DetailPresenter.java
package com.test.upcoming.ui.activity.detail;

import com.test.upcoming.model.MovieImages;
import com.test.upcoming.networkcall.WebService;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by Aks4125 on 7/26/2017.
 */

public class DetailPresenter implements DetailContractor.IDetailPresenter {

    private DetailContractor.IDetailView detailView;
    private WebService webService;
    private CompositeDisposable mDisposable;

    @Inject
    public DetailPresenter(WebService webService) {
        this.webService = webService;
    }

    public void setDetailView(DetailContractor.IDetailView detailView) {
        this.detailView = detailView;
    }

    private void onError(Throwable throwable) {
        unsubscribe();// dispose the disposable
        detailView.showMessage("Something went wrong");
        detailView.stopProgress();
    }

    private void onSuccess(MovieImages movieImages) {
        unsubscribe();
        detailView.stopProgress();
        detailView.processJson(movieImages);
    }

    @Override
    public void subscribe() {
        mDisposable = new CompositeDisposable();
    }


    @Override
    public void getMovieImages(String apiKey, String movieID) {
        detailView.showProgress();
        subscribe();
        mDisposable.add(
                webService.getMovieImages(movieID, apiKey)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(this::onSuccess, this::onError)

        );
    }

    @Override
    public void unsubscribe() {
        if (mDisposable != null && !mDisposable.isDisposed())
            mDisposable.dispose();
        detailView.stopProgress();
    }


}

File: app/src/main/java/com/test/upcoming/ui/activity/detail/DetailPresenter.kt
package com.test.upcoming.ui.activity.detail

import com.test.upcoming.networkcall.WebService
import com.google.gson.JsonObject

import javax.inject.Inject

import rx.Subscriber
import rx.Subscription
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers

/**
 * Created by Aks4125 on 7/26/2017.
 */

class DetailPresenter @Inject
constructor(private val webService: WebService) : DetailContractor.IDetailPresenter {

    private var detailView: DetailContractor.IDetailView? = null
    private var mSubscription: Subscription? = null

    fun setDetailView(detailView: DetailContractor.IDetailView) {
        this.detailView = detailView
    }


    override fun subscribe() {

    }

    override fun getMovieImages(apiKey: String, movieID: String) {
        detailView!!.showProgress()

        mSubscription = webService.getMovieImages(movieID, apiKey)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Subscriber<JsonObject>() {
                    override fun onCompleted() {

                    }

                    override fun onError(e: Throwable) {
                        detailView!!.showMessage("Something went wrong")
                        detailView!!.stopProgress()
                    }

                    override fun onNext(jsonObject: JsonObject) {
                        detailView!!.stopProgress()
                        detailView!!.processJson(jsonObject)
                    }
                })
    }

    override fun unsubscribe() {
        if (mSubscription != null && !mSubscription!!.isUnsubscribed)
            mSubscription!!.unsubscribe()

        detailView!!.stopProgress()
    }
}
